File: gns3-net-proxy.py
#!/usr/bin/env python3

import os
import argparse
import fcntl
import struct
import select
import subprocess
import sys
import threading
import time
import traceback

def open_tap(name):
    TUNSETIFF = 0x400454ca
    IFF_TAP = 0x0002
    IFF_NO_PI = 0x1000
    
    # Solo intentamos abrir una vez. Si falla, es fatal porque la interfaz
    # YA DEBERÍA EXISTIR gracias a docker_vm.py
    try:
        print(f"DEBUG: Opening /dev/net/tun for attaching to '{name}'...", file=sys.stderr)
        tun = os.open('/dev/net/tun', os.O_RDWR)
        
        # Preparamos el struct para TUNSETIFF
        # Esto le dice al Kernel: "Quiero controlar la interfaz 'name' existente"
        ifr = struct.pack('16sH', name.encode()[:15], IFF_TAP | IFF_NO_PI)
        
        # Adjuntamos el file descriptor a la interfaz
        fcntl.ioctl(tun, TUNSETIFF, ifr)
        
        print(f"DEBUG: Successfully attached to TAP '{name}'", file=sys.stderr)
        
        # --- SECCIÓN ELIMINADA ---
        # No intentamos ponerla UP aquí. Eso se hace en el host con 'ip link set up'.
        # Intentar hacerlo aquí con el fd del tun genera Errno 22.
        # -------------------------
        
        return tun
            
    except Exception as e:
        print(f"ERROR attaching to TAP {name}: {e}", file=sys.stderr)
        traceback.print_exc(file=sys.stderr)
        sys.exit(1)
        
def read_stderr(proc, cid):
    """Lee stderr del agente y lo imprime"""
    print(f"DEBUG: Starting stderr reader thread for container {cid}", file=sys.stderr)
    try:
        while True:
            line = proc.stderr.readline()
            if line:
                print(f"[Agent@{cid}] {line.decode()}", file=sys.stderr, end='')
            else:
                print(f"DEBUG: Agent stderr closed for {cid}", file=sys.stderr)
                break
    except Exception as e:
        print(f"ERROR in stderr reader: {e}", file=sys.stderr)

# Redirigir stderr a un archivo para depurar
sys.stderr = open('/tmp/gns3_proxy_debug.log', 'a')

def main():
    p = argparse.ArgumentParser()
    p.add_argument('--tap', required=True)
    p.add_argument('--cid', required=True, help="Container ID")
    p.add_argument('--agent-path', required=True, help="Path to agent binary inside container")
    p.add_argument('--ifname', required=True, help="Interface name inside container")
    p.add_argument('--mac', required=False, help="MAC address")
    args = p.parse_args()

    print(f"=== GNS3 Proxy Starting ===", file=sys.stderr)
    print(f"Timestamp: {time.ctime()}", file=sys.stderr)
    print(f"TAP: {args.tap}", file=sys.stderr)
    print(f"Container: {args.cid}", file=sys.stderr)
    print(f"Agent: {args.agent_path}", file=sys.stderr)
    print(f"Interface: {args.ifname}", file=sys.stderr)
    print(f"MAC: {args.mac}", file=sys.stderr)
    print(f"Python version: {sys.version}", file=sys.stderr)
    print(f"Current user: {os.getuid()}/{os.geteuid()}", file=sys.stderr)

    # 1. Abrir TAP
    print(f"\nDEBUG: Step 1 - Opening TAP device...", file=sys.stderr)
    tap_fd = open_tap(args.tap)
    print(f"DEBUG: TAP {args.tap} opened (fd={tap_fd})", file=sys.stderr)

    # 2. Construir comando para el AGENTE BINARIO
    print(f"\nDEBUG: Step 2 - Building agent command...", file=sys.stderr)
    cmd = [
        'podman', 'exec', '-i', args.cid,
        args.agent_path,
        '--ifname', args.ifname
    ]
    if args.mac:
        cmd.extend(['--mac', args.mac])

    print(f"DEBUG: Full command: {' '.join(cmd)}", file=sys.stderr)
    
    # Verificar si el contenedor existe
    print(f"DEBUG: Checking if container {args.cid} exists...", file=sys.stderr)
    check_cmd = ['podman', 'ps', '-q', '--filter', f'id={args.cid}']
    result = subprocess.run(check_cmd, capture_output=True, text=True)
    if result.returncode != 0 or not result.stdout.strip():
        print(f"ERROR: Container {args.cid} not found or not running!", file=sys.stderr)
        print(f"DEBUG: podman ps output: {result.stdout}", file=sys.stderr)
        os.close(tap_fd)
        sys.exit(1)
    print(f"DEBUG: Container {args.cid} is running", file=sys.stderr)

    # 3. Lanzar el proceso del agente
    print(f"\nDEBUG: Step 3 - Launching agent process...", file=sys.stderr)
    try:
        proc = subprocess.Popen(cmd, 
                              stdin=subprocess.PIPE, 
                              stdout=subprocess.PIPE, 
                              stderr=subprocess.PIPE,
                              bufsize=0)
        print(f"DEBUG: Agent process started with PID {proc.pid}", file=sys.stderr)
        
        # Esperar un momento para ver si el agente muere inmediatamente
        time.sleep(0.5)
        if proc.poll() is not None:
            print(f"ERROR: Agent process died immediately with exit code {proc.returncode}", file=sys.stderr)
            # Leer cualquier salida del agente
            stdout, stderr = proc.communicate()
            print(f"DEBUG: Agent stdout: {stdout[:200]}", file=sys.stderr)
            print(f"DEBUG: Agent stderr: {stderr[:200]}", file=sys.stderr)
            os.close(tap_fd)
            sys.exit(1)
            
    except Exception as e:
        print(f"ERROR: Failed to start agent process: {e}", file=sys.stderr)
        traceback.print_exc(file=sys.stderr)
        os.close(tap_fd)
        sys.exit(1)

    # Hilo para leer stderr del agente
    print(f"DEBUG: Starting stderr reader thread...", file=sys.stderr)
    stderr_thread = threading.Thread(target=read_stderr, args=(proc, args.cid[:8]), daemon=True)
    stderr_thread.start()

    print(f"\nDEBUG: Proxy bridge initialized: {args.tap} <-> {args.cid}:{args.ifname}", file=sys.stderr)
    print(f"DEBUG: Entering main bridge loop...", file=sys.stderr)

    # 4. Bucle de retransmisión
    loop_count = 0
    try:
        while True:
            loop_count += 1
            if loop_count % 100 == 0:
                print(f"DEBUG: Bridge loop iteration {loop_count}", file=sys.stderr)
            
            readable, _, _ = select.select([tap_fd, proc.stdout], [], [], 1.0)
            
            if not readable:
                # Timeout - verificar que el agente sigue vivo
                if proc.poll() is not None:
                    print(f"ERROR: Agent process exited with code {proc.returncode}", file=sys.stderr)
                    break
                continue

            if tap_fd in readable:
                data = os.read(tap_fd, 65535)
                if not data: 
                    print("ERROR: TAP device closed", file=sys.stderr)
                    break
                print(f"DEBUG: Read {len(data)} bytes from TAP", file=sys.stderr)
                try:
                    written = proc.stdin.write(data)
                    proc.stdin.flush()
                    print(f"DEBUG: Wrote {written} bytes to agent stdin", file=sys.stderr)
                except BrokenPipeError:
                    print("ERROR: Agent stdin broken pipe", file=sys.stderr)
                    break

            if proc.stdout in readable:
                data = os.read(proc.stdout.fileno(), 65535)
                if not data: 
                    print("ERROR: Agent stdout closed", file=sys.stderr)
                    break
                print(f"DEBUG: Read {len(data)} bytes from agent stdout", file=sys.stderr)
                written = os.write(tap_fd, data)
                print(f"DEBUG: Wrote {written} bytes to TAP", file=sys.stderr)
                
            if proc.poll() is not None:
                print(f"ERROR: Agent process exited with code {proc.returncode}", file=sys.stderr)
                break

    except KeyboardInterrupt:
        print("\nDEBUG: Proxy interrupted by user", file=sys.stderr)
    except Exception as e:
        print(f"ERROR in bridge loop: {e}", file=sys.stderr)
        traceback.print_exc(file=sys.stderr)
    finally:
        print("\nDEBUG: Cleaning up...", file=sys.stderr)
        if proc.poll() is None:
            print("DEBUG: Terminating agent process...", file=sys.stderr)
            proc.terminate()
            try:
                proc.wait(timeout=2)
            except subprocess.TimeoutExpired:
                print("DEBUG: Agent did not terminate, killing...", file=sys.stderr)
                proc.kill()
        os.close(tap_fd)
        print("DEBUG: Proxy shutdown complete", file=sys.stderr)

if __name__ == '__main__':
    main()

File: gns3-net-proxy.cpp
/*
 * GNS3 Proxy - C++ Implementation
 * Replaces the Python script for lower latency and CPU usage.
 * Compile with: g++ -O3 -pthread -o gns3-net-proxy gns3-net-proxy.cpp
 */

#include <iostream>
#include <vector>
#include <string>
#include <thread>
#include <cstring>
#include <csignal>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <linux/if.h>
#include <linux/if_tun.h>
#include <sys/wait.h>
#include <sys/poll.h>
#include <chrono>
#include <ctime>


// Buffer configuration (same as in Python 65535)
constexpr size_t BUFFER_SIZE = 65535;
volatile std::sig_atomic_t g_running = 1;


// Structure for arguments
struct Config {
    std::string tap_dev;
    std::string cid;
    std::string agent_path;
    std::string ifname;
    std::string mac;
};

// Signal handler (Ctrl+C)
void signal_handler(int signal) {
    if (signal == SIGINT || signal == SIGTERM) {
        g_running = 0;
    }
}

// Function to open the TAP interface
int open_tap(const std::string& name) {
    std::cerr << "DEBUG: Opening /dev/net/tun for attaching to '" << name << "'..." << std::endl;

    int fd = open("/dev/net/tun", O_RDWR);
    if (fd < 0) {
        perror("ERROR opening /dev/net/tun");
        exit(1);
    }

    struct ifreq ifr;
    std::memset(&ifr, 0, sizeof(ifr));
    ifr.ifr_flags = IFF_TAP | IFF_NO_PI;
    std::strncpy(ifr.ifr_name, name.c_str(), IFNAMSIZ);

    if (ioctl(fd, TUNSETIFF, (void*)&ifr) < 0) {
        perror(("ERROR attaching to TAP " + name).c_str());
        close(fd);
        exit(1);
    }

    std::cerr << "DEBUG: Successfully attached to TAP '" << name << "'" << std::endl;
    return fd;
}

// Function to check if the container exists
void check_container(const std::string& cid) {
    std::string cmd = "podman ps -q --filter id=" + cid;
    FILE* pipe = popen(cmd.c_str(), "r");
    if (!pipe) {
        std::cerr << "ERROR: Failed to run podman check." << std::endl;
        exit(1);
    }
    char buffer[128];
    std::string result = "";
    while (fgets(buffer, sizeof(buffer), pipe) != NULL) {
        result += buffer;
    }
    pclose(pipe);

    if (result.empty()) {
        std::cerr << "ERROR: Container " << cid << " not found or not running!" << std::endl;
        exit(1);
    }
    std::cerr << "DEBUG: Container " << cid << " is running" << std::endl;
}

// Thread to read stderr from the agent
void read_stderr_thread(int fd, std::string cid) {
    std::cerr << "DEBUG: Starting stderr reader thread for container " << cid << std::endl;
    char buffer[1024];
    ssize_t bytes_read;
    while ((bytes_read = read(fd, buffer, sizeof(buffer) - 1)) > 0) {
        buffer[bytes_read] = '\0';
        std::cerr << "[Agent@" << cid << "] " << buffer;
    }
    std::cerr << "DEBUG: Agent stderr closed for " << cid << std::endl;
}

int main(int argc, char* argv[]) {
    Config config;

    // Simple manual parsing of arguments
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "--tap" && i + 1 < argc) config.tap_dev = argv[++i];
        else if (arg == "--cid" && i + 1 < argc) config.cid = argv[++i];
        else if (arg == "--agent-path" && i + 1 < argc) config.agent_path = argv[++i];
        else if (arg == "--ifname" && i + 1 < argc) config.ifname = argv[++i];
        else if (arg == "--mac" && i + 1 < argc) config.mac = argv[++i];
    }

    if (config.tap_dev.empty() || config.cid.empty() || config.agent_path.empty() || config.ifname.empty()) {
        std::cerr << "Usage: " << argv[0] << " --tap <name> --cid <id> --agent-path <path> --ifname <name> [--mac <mac>]" << std::endl;
        return 1;
    }

    // freopen("/tmp/gns3_proxy_cpp_debug.log", "a", stderr);
    auto now = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
    std::cerr << "=== GNS3 Proxy Starting (C++) ===" << std::endl;
    std::cerr << "Timestamp: " << std::ctime(&now);
    std::cerr << "TAP: " << config.tap_dev << std::endl;
    std::cerr << "Container: " << config.cid << std::endl;

    // 1. Open TAP
    int tap_fd = open_tap(config.tap_dev);

    // 2. Verify Container
    check_container(config.cid);

    // 3. Prepare Pipes for the subprocess
    int pipe_stdin[2], pipe_stdout[2], pipe_stderr[2];
    if (pipe(pipe_stdin) < 0 || pipe(pipe_stdout) < 0 || pipe(pipe_stderr) < 0) {
        perror("ERROR creating pipes");
        return 1;
    }

    // 4. Launch agent process (Fork + Exec)
    pid_t pid = fork();
    if (pid == 0) {
        // Redirigir stdin/stdout/stderr
        dup2(pipe_stdin[0], STDIN_FILENO);
        dup2(pipe_stdout[1], STDOUT_FILENO);
        dup2(pipe_stderr[1], STDERR_FILENO);

        // Close unused ends in the child
        close(pipe_stdin[1]);
        close(pipe_stdout[0]);
        close(pipe_stderr[0]);
        close(tap_fd); // The child does not need the TAP.

        // Build arguments for execvp
        std::vector<char*> args;
        args.push_back(strdup("podman"));
        args.push_back(strdup("exec"));
        args.push_back(strdup("-i"));
        args.push_back(strdup(config.cid.c_str()));
        args.push_back(strdup(config.agent_path.c_str()));
        args.push_back(strdup("--ifname"));
        args.push_back(strdup(config.ifname.c_str()));

        if (!config.mac.empty()) {
            args.push_back(strdup("--mac"));
            args.push_back(strdup(config.mac.c_str()));
        }
        args.push_back(nullptr);

        execvp("podman", args.data());

        perror("ERROR executing podman");
        exit(1);
    } else if (pid < 0) {
        perror("ERROR forking");
        return 1;
    }

    // --- PARENT PROCESS (Proxy) ---
    
    // Close pipe ends used by the child
    close(pipe_stdin[0]);
    close(pipe_stdout[1]);
    close(pipe_stderr[1]);

    int agent_in_fd = pipe_stdin[1];   // We write here (goes to the agent's stdin)
    int agent_out_fd = pipe_stdout[0]; // We read from here (comes from the agent's stdout)
    int agent_err_fd = pipe_stderr[0]; // We read stderr

    std::cerr << "DEBUG: Agent process started with PID " << pid << std::endl;

    // Thread for stderr
    std::thread stderr_thread(read_stderr_thread, agent_err_fd, config.cid.substr(0, 8));
    stderr_thread.detach();

    // Configure signals
    std::signal(SIGINT, signal_handler);
    std::signal(SIGTERM, signal_handler);

    // 5. Main Loop (Bridge Loop)
    // We use poll() instead of select() because it is more modern and efficient in C++
    struct pollfd fds[2];
    fds[0].fd = tap_fd;
    fds[0].events = POLLIN; // Listen to TAP data
    fds[1].fd = agent_out_fd;
    fds[1].events = POLLIN; // Listen to Agent data

    std::vector<char> buffer(BUFFER_SIZE);

    std::cerr << "DEBUG: Entering main bridge loop..." << std::endl;

    while (g_running) {
        // 1000ms timeout to check if g_running changed or the child died
        int ret = poll(fds, 2, 1000); 

        if (ret < 0) {
            if (errno == EINTR) continue; // Interrupted by signal
            perror("ERROR in poll");
            break;
        }

        // Check if the child process is still alive
        int status;
        if (waitpid(pid, &status, WNOHANG) != 0) {
            std::cerr << "ERROR: Agent process exited unexpectedly." << std::endl;
            g_running = 0;
            break;
        }

        if (ret == 0) continue; // Timeout

        // --- DATA FROM THE TAP -> AGENT ---
        if (fds[0].revents & POLLIN) {
            ssize_t n = read(tap_fd, buffer.data(), BUFFER_SIZE);
            if (n <= 0) break; // Error or shutdown

            // Write everything to the agent's stdin
            ssize_t written = 0;
            while (written < n) {
                ssize_t w = write(agent_in_fd, buffer.data() + written, n - written);
                if (w < 0) {
                    perror("ERROR writing to agent stdin");
                    g_running = 0;
                    break;
                }
                written += w;
            }
        }

        // --- DATA FROM AGENT -> TAP ---
        if (fds[1].revents & POLLIN) {
            ssize_t n = read(agent_out_fd, buffer.data(), BUFFER_SIZE);
            if (n <= 0) break; // EOF o error

            // Write everything to the TAP
            ssize_t written = 0;
            while (written < n) {
                ssize_t w = write(tap_fd, buffer.data() + written, n - written);
                if (w < 0) {
                    perror("ERROR writing to TAP");
                    g_running = 0;
                    break;
                }
                written += w;
            }
        }
    }

    std::cerr << "\nDEBUG: Cleaning up..." << std::endl;
    close(tap_fd);
    close(agent_in_fd);
    close(agent_out_fd);
    close(agent_err_fd);

    // Ensure that we kill the child process
    kill(pid, SIGTERM);
    waitpid(pid, nullptr, 0);

    std::cerr << "DEBUG: Proxy shutdown complete" << std::endl;
    return 0;
}
